### PR TITLE
Preventing author profile from going transparent

### LINKED ISSUE #15290: Hover card of post author doesn't looks well styled
**Describe the bug**

The hover card of post author doesn't looks well when it is near the bottom of browser window. I think, it breaks down due to insufficient space.

**To Reproduce**

1. Go to 'dev.to' (or any other website which uses Forem)
2. Scroll down and stop on any post author near bottom of browser window
3. Hover on 'post author name'
4. See bug

**Expected behavior**

The hover card of post author should looks like as expected.

**Screenshots**

screenshot-hovercard

**Desktop:**

- OS, version: MacOS Monterey, 12.0.1
- Browser, version: Google Chrome, 95.0.4638.69

**Additional Context:**

Also here: (https://community.devocate.com/)

https://user-images.githubusercontent.com/41960283/140640366-bdda99d4-296a-4218-b151-4fb0d2edb9d4.mov

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [x] Bug Fix

## Description

Prior to this change, when you would hover over an author profile that
was "close to" the bottom of the page, the card would be transparent
instead of the expected opaque.

With this change, when you hover over an author profile that is "close"
to the bottom of the page, the card renders as an opaque card.

Now why does this work?  I don't really know, except to say that a
`display: table` imperative is a more "chunky" display element than a
block.  I would love for someone to explain why this works.

Tested against on MacOS Safari, Firefox, and Chrome.

https://developer.mozilla.org/en-US/docs/Web/CSS/display

## Related Tickets & Documents

Fixes #15290

## QA Instructions, Screenshots, Recordings

Open to the homepage, and scroll so that the bottom-most visible article card is position just above the browser window.  Hover over the author.  With this fix, the card should be opaque and rendered past the bottom of the view frame.

Without this commit, the card will be transparent-ish and the text from the article card will "bleed" through.

### UI accessibility concerns?

Uncertain.

## Added/updated tests?

- [x] No, and this is why: I don't believe it's necessary.

## [Forem core team only] How will this change be communicated?

- [x] This change does not need to be communicated, and this is why not: it is a small bug fix for a UI edge case (albeit not an extreme edge case but a somewhat common one).

### CODE REVIEW

**Review by ludwiczakpawel — CHANGES_REQUESTED** [review]
So here's what happens _usually_ (but not for these profile-preview dropdowns): dropdown component is being attached to the bottom element of its trigger unless the trigger (or the dropdown, I'm not sure) is close to the bottom edge of the viewport - in that case the dropdown changes its position and appears ABOVE the trigger (see "Subscribe" component on top of the Discussion section on a post view for a properly working example). Close enough to be actually "cut off".  That behavior is achieved by applying [`reverse` class](https://github.com/forem/forem/blob/e7af2bfc86d52322d79d4186cb4cc69c4e8ddd92/app/assets/stylesheets/components/dropdowns.scss#L20-L24) which happens with JS.

Why it doesn't work for these particular dropdowns? Because of [classes specificity](https://www.w3schools.com/css/css_specificity.asp) - we have [applied some custom styling](https://github.com/forem/forem/blob/main/app/assets/stylesheets/components/profile-preview-card.scss#L10) to these specific (profile preview) dropdowns so they became more important than the ones we defined on [the component level](https://github.com/forem/forem/blob/main/app/assets/stylesheets/components/dropdowns.scss#L20-L24).

I think there was a reason why we added that custom styling but I'm not 100% sure why exactly. Maybe @nickytonline or @aitchiss know.

Even though the `display: table` may fix the bug at first glance, (which btw. really surprised me! :D) it doesn't fix the core issue and it certainly doesn't bring bring the `revert` functionality.

I think there are two solutions:

1. Change the specificity of that custom CSS so it doesn't overwrite the component CSS - but this may not be trivial...
2. Add `!important`s to [the component's `revert` styling](https://github.com/forem/forem/blob/main/app/assets/stylesheets/components/dropdowns.scss#L20-L24) - even though we tend to avoid using `!important` in our CSS, I think using it here is justified.

**Comment by aitchiss** [comment]
> I think there was a reason why we added that custom styling but I'm not 100% sure why exactly. Maybe @nickytonline or @aitchiss know.

To be honest I can't remember 😆  I'm happy to dig into it and refresh my memory though, just let me know if I can help. It's entirely possible that we don't need that level of specificity and it was just an oversight.

I'm not sure why the `top` styling and `reverse` class would affect the transparency of the background though? As far as I understand, the card is being positioned properly, but the background becomes transparent

**Comment by aitchiss** [comment]
For what it's worth, I think the root cause of this issue will be in the listener which repositions the dropdown - looks like it's not functioning properly / could be refactored to use the `reverse` class with a bit of investigation: https://github.com/forem/forem/blob/86e4c39e31ddae896e8f70581b73435b25d935cd/app/javascript/utilities/dropdownUtils.js#L22

File: app/assets/stylesheets/components/dropdowns.scss [hunk 1]
```diff
@@ -2,9 +2,11 @@
 
 .crayons-dropdown {
   position: absolute;
+  top: 100%;
```
Comment (ludwiczakpawel) [inline app/assets/stylesheets/components/dropdowns.scss:5]: Turns out most of our dropdown are positioned this way so it made sense to have it there.

File: app/assets/stylesheets/components/dropdowns.scss [hunk 2]
```diff
@@ -2,9 +2,11 @@
...
   display: none; // initially hidden
   padding: var(--su-2);
   min-width: 250px;
+  margin-top: var(--su-1);
```
Comment (ludwiczakpawel) [inline app/assets/stylesheets/components/dropdowns.scss:9]: Not really part of this PR but I wanted to sneak it in.

File: app/assets/stylesheets/components/profile-preview-card.scss [hunk 3]
```diff
@@ -24,7 +23,7 @@
 
   &__trigger:hover + .profile-preview-card__content.crayons-dropdown {
     display: block;
-    animation: hoverAppear vaR(--dropdown-hover-delay);
+    animation: hoverAppear var(--dropdown-hover-delay);
```
Comment (ludwiczakpawel) [inline app/assets/stylesheets/components/profile-preview-card.scss:26]: i couldn't stand it 🤷‍♂️  :D

File: app/javascript/CommentSubscription/CommentSubscription.jsx [hunk 4]
```diff
@@ -38,39 +38,6 @@ export class CommentSubscription extends Component {
...
-
-    if (
-      Math.sign(dropDownBottom) === -1 ||
-      dropDownBottom > window.innerHeight
-    ) {
-      // The 4 pixels is the box shadow from the drop down.
-      element.style.bottom = `${height + 4}px`;
-    }
-  };
-
```
Comment (ludwiczakpawel) [inline app/javascript/CommentSubscription/CommentSubscription.jsx:73]: So one of our dropdowns (for Discussion's _Subscribe_ button) was actually doing this repositioning but it was completely custom for some reason (I guess it was created BEFORE the `reverse` magic). So I removed it and updated the component itself with `reverse` magic handler.

File: app/views/articles/_actions.html.erb [hunk 5]
```diff
@@ -33,7 +33,7 @@
         <%= inline_svg_tag("overflow-horizontal.svg", aria_hidden: true, class: "dropdown-icon crayons-icon", title: t("views.actions.more.title")) %>
       </button>
 
-      <div id="article-show-more-dropdown" class="crayons-dropdown side-bar left-2 right-2 m:right-auto m:left-100 s:left-auto bottom-100 m:bottom-auto m:top-0 mb-1 m:mb-0">
```
Comment (ludwiczakpawel) [inline app/views/articles/_actions.html.erb:36]: little fix and cleanup

**Comment by ludwiczakpawel** [comment]
After chatting with @aitchiss we decided to do a little cleanup. And apparently there was a bit more refactor needed 🤷... I left some comments in the code. Anyway, things should now work properly.

**Review by ludwiczakpawel — APPROVED** [review]
I initially requested changes, but I have also pushed some changes afterwards... And I can now only "approve" it (I can't take back my request changes), so extra two reviews would be appreciated 🤷‍♂️ 

image

File: app/assets/stylesheets/components/profile-preview-card.scss (hunk 3 above)
Comment (aitchiss) [reply app/assets/stylesheets/components/profile-preview-card.scss:26]: hahaha woops

**Review by aitchiss — APPROVED** [review]
This all looks good to me - nice one!

**Review by fdoxyz — APPROVED** [review]
Tried it out locally and seems like it works great for me 🙌

**Review by msarit — APPROVED** [review]
All changes made appear to work well! I replicated the bug on DEV; this PR fixes the issue, and also appears to shift the pop-up profile higher when it renders close to the browser-edge.

Great work all around!

**Pull request merged**